### PR TITLE
[WIP] Improved CPU bus implementation

### PR DESCRIPTION
**Pull request by Wolfvak**
TODO:

- Test absolutely everything

- Make sure everything is endian-independent (not completely sure how C# handles those cases)

- Move ITCM/DTCM to the ARM9 core and the MPCore PMR to the ARM11 core

- Add WriteUInt16/WriteUInt8 operations that don't rely on RMW

- Better memory mapped peripheral support (parent bus, update on tick, etc)

- Anything else I forgot about

### CODE REVIEW

File: LemonLime.LLE/CPUBus.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            CPUMap Map = FindMap(Address);
+            return Map.Device.ReadUInt32(Address - Map.Start);
+        }
+
+        public ushort ReadUInt16(uint Address)
+        {
+            uint Mask = Address & 2;
+            uint MaskAddr = Address - Mask;
+            return (ushort)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr)*8));
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs]: Code styling: Please make `return (ushort)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr)*8));` `return (ushort)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr) * 8));`

File: LemonLime.LLE/CPUBus.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        public byte ReadUInt8(uint Address)
+        {
+            uint Mask = Address & 3;
+            uint MaskAddr = Address - Mask;
+            return (byte)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr)*8));
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs]: Code styling: Please make `return (byte)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr)*8));` `return (byte)(ReadUInt32(MaskAddr) >> (int)((Address - MaskAddr) * 8));`

File: LemonLime.LLE/CPUBus.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        public void Attach(CPUDevice Device, uint Start, uint End)
+        {
+            BusMap.Add(new CPUMap(Device, Start, End));
+        }
+
+        private CPUMap FindMap(uint Address)
+        {
+            CPUMap Map = BusMap.Where(map => Address >= map.Start && Address <= map.End).SingleOrDefault();
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs]: Code styling: Make `map` `FMap`

File: LemonLime.LLE/CPUBus.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    class CPUBus : IBus
+    {
+        class CPUMap
+        {
+            public CPUDevice Device;
+            public uint Start, End;
+            public CPUMap(CPUDevice Device, uint Start, uint End)
+            {
+                this.Device = Device;
+                this.Start = Start;
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs:20]: Make this `Base`?

File: LemonLime.LLE/CPUBus.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                this.End = End;
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs]: Make this `Size`?

File: LemonLime.LLE/CPUBus.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+using System.Linq;
+
+namespace LemonLime.LLE
+{
+    class CPUBus : IBus
+    {
+        class CPUMap
+        {
+            public CPUDevice Device;
+            public uint Start, End;
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUBus.cs:15]: Code styling: Space between public definition

File: LemonLime.LLE/CPUDevice.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    public interface CPUDevice
+    {
+        uint ReadUInt32(uint Offset);
+        void WriteUInt32(uint Offset, uint Word);
+        void Update();
+    }
+
+    class CPURAM : CPUDevice
+    {
+    	private uint[] RAM;
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUDevice.cs]: Code styling: Space between public definition

File: LemonLime.LLE/CPUDevice.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    	{
+    		this.RAM[Offset / 4] = Word;
+    	}
+
+    	public void Update() {}
+    }
+
+    class CPUROM : CPUDevice
+    {
+    	private byte[] ROM;
```
Comment (dd01da0465b4542f8f8af4ecedc149ed) [inline LemonLime.LLE/CPUDevice.cs]: Code styling: Space between public definition

**Review by dd01da0465b4542f8f8af4ecedc149ed — CHANGES_REQUESTED** [review]

File: LemonLime.LLE/CPUBus.cs (hunk 5 above)
Comment (Wolfvak) [reply LemonLime.LLE/CPUBus.cs]: No, `Size` implies, well, size. These variables indicate a range, so `Start` and `End` are more accurate, at least more than `Size`.

Comment (dd01da0465b4542f8f8af4ecedc149ed) [reply LemonLime.LLE/CPUBus.cs]: Ah, alright!

**Comment by dd01da0465b4542f8f8af4ecedc149ed** [comment]
Might wanna make the 2 spaces 4 spaces in the code :P

**Comment by Wolfvak** [comment]
Yeah, Sublime is being dumb and doesn't know whether it should default to spaces or tabs for some reason, so everything looks off when pushed. Will fix later, but right now the technical aspect takes priority over anything else.

About the implementation of WriteUInt8/WriteUInt16, I'm having some second thoughts. On one hand you get the core running faster because there's no need for Read-Modify-Write and the hardware emulation becomes more flexible (very important with some peripherals), but it can potentially lead to a lot of code repetition.

I'll implement it anyway, just putting my thoughts out there.

**Review by dd01da0465b4542f8f8af4ecedc149ed — APPROVED** [review]

**Pull request merged**